google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutations.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.models;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.InternalApi;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.common.base.Preconditions;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import org.threeten.bp.Duration;
+
+/**
+ * Tracker for outstanding bulk mutations. Allows for the caller to wait for all of the outstanding
+ * mutations to complete.
+ *
+ * @see com.google.cloud.bigtable.data.v2.BigtableDataClient#newBulkMutations() for example usage.
+ */
+public final class BulkMutations implements AutoCloseable {
```

Comment:
I get the feeling that this could be represented as a callable somehow - it seems to function like a client streaming callable.

Reply:
I don't think it fits as either:
- It's not quite a UnaryCallable because a sync `call()` doesn't make sense
- It's not quite a ClientStreamingCallable because each `send` returns a future, while client streaming callables don't make any assumptions about responses.

Reply:
Yes, you're correct about the mismatch with `ClientStreamingCallable`, which only has one response at the end.

Since this is making up for deficiencies in batching, let's try to match the naming of batching. I would suggest:

- `BulkMutations` -> `BulkMutationBatcher`
- `send()` -> `add()`
- `bigtableDataClient.newBulkMutations` -> `bigtableDataClient.newBulkMutationBatcher`

Since this might be replaced by a parameterized class in gax, let's mark it `@BetaApi` even though this library is still alpha. Maybe add text in the annotation like "This surface is likely to change as the batching surface evolves."